### ISSUE TITLE
Full Text Search: Matching a phrase using only a few letters

### ISSUE THREAD

**Issue by EdIzaguirre** [groue/GRDB.swift#908]
### What did you do?
Hi Groue,

I am attempting to use FTS5 to find documents in a virtual table that match a pattern. The relevant code is below:

              try dbQueue.read { db in
                let pattern = FTS5Pattern(matchingPrefixPhrase: filteredText)

                let matchGroups = try Group // fetches [Group]
                    .matching(pattern)
                    .order(Column.rank)
                    .fetchAll(db)

                print(matchGroups)

                filteredGroups = [Group]()
                for group in matchGroups{
                    filteredGroups.append(group)
                }

                if searchText == "" {
                    filteredGroups = groups!
                }
                tableView.reloadData()
            }

### What did you expect to happen?
I expected that when a user begins typing, for example the letters 'Dow' into the search bar, a group with the name "Downey Soccer Club" would be considered a match.
 
### What happened instead?
The group is only considered a match when the entire word is spelled out. For example, the user must type in "Downey" in order for the "Downey Soccer Club" group to be considered a match. Looking through the documentation, I thought that matchingPrefixPhrase would accomplish this, but it did not. 

It would be even better if a user could type in "owne" into the search bar and the "Downey Soccer Club" would be considered a match. Looking through SQLite documentation, I see that I may need the "LIKE" token instead of the "MATCH to accomplish this. Does this seem accurate, and if so is there a way to do this in GRDB? (I know SQLite can be used in GRDB, but I do seem to run into problems doing this, so was hoping for a swift way instead :)

Thank you again.

**Comment by groue** [comment]
Hello @EdIzaguirre,

The documentation could be improved indeed, and it will. `FTS5Pattern(matchingPrefixPhrase:)` looks for *full tokens at the beginning of the document*. Given the `the beautiful rainbow` document:

- `the` matches
- `the beautiful` matches
- `the beau` does not match
- `beautiful` does not match
- `th` does not match
- `beau` does not match

> It would be even better if a user could type in "owne" into the search bar and the "Downey Soccer Club" would be considered a match.

This is not how FTS5 works. You need LIKE for that indeed (Swift version documented under [SQL Operators](https://github.com/groue/GRDB.swift/blob/master/README.md#sql-operators)).

Don't give up: it takes time to figure out exactly what an app needs, and how to leverage SQLite for it. Trials and errors are often part of the process!

**Comment by EdIzaguirre** [comment]
Thanks as always for a quick response! I was able to use LIKE to find the appropriate groups. However, I think sticking to FTS5 for the bm25 ranking is too good to pass up.

Last questions: I see on the [SQLite Documentation](https://www.sqlite.org/fts5.html) under "4.3.4. The Experimental Trigram Tokenizer" that there is an experimental tokenizer that could do what I want. However under the tokenizers for GRDB I only see the standard Unicode, ASCII, and Porter. Any chance the Trigram is also in GRDB, or if I can just use SQLite for it?

Also, under the section "3.3. FTS5 Prefix Queries" I see that the '*' wildcard can be used to denote a prefix token. However, adding '*' to the end of the user's searchText string fails to find any match at all. Is this something that must be handled with SQLite, or can I do this in Swift?

**Comment by groue** [comment]
> Last questions: I see on the SQLite Documentation under "4.3.4. The Experimental Trigram Tokenizer" that there is an experimental tokenizer that could do what I want. However under the tokenizers for GRDB I only see the standard Unicode, ASCII, and Porter. Any chance the Trigram is also in GRDB, or if I can just use SQLite for it?

I did not know about it :-) Trigram was shipped with the latest SQLite 3.34.0 (see https://sqlite.org/changes.html). No Apple operating system ships with it today: you'll need a custom SQLite build. And then:

```swift
try db.create(virtualTable: "book", using: FTS5()) { t in
    t.tokenizer = FTS5TokenizerDescriptor(components: ["trigram"])
    ...
}
```

> Also, under the section "3.3. FTS5 Prefix Queries" I see that the '*' wildcard can be used to denote a prefix token. However, adding '*' to the end of the user's searchText string fails to find any match at all. Is this something that must be handled with SQLite, or can I do this in Swift?

Thanks for taking the time to look at the SQLite doc.

Currently, GRDB builds [FTS5Pattern](https://github.com/groue/GRDB.swift/blob/master/Documentation/FullTextSearch.md#fts5pattern) from a **raw pattern** (`Database.makeFTS5Pattern(rawPattern:forTable:)`), or from **user-provided strings**, such as strings typed in a text field (the various failable FTS5Pattern initializers).

The patterns built from user-provided strings deliberately ignore all special query syntax. You can try it yourself:

```swift
FTS5Pattern(matchingAnyTokenIn: "foo bar")?.rawPattern    // foo OR bar
FTS5Pattern(matchingAllTokensIn: "foo bar")?.rawPattern   // foo bar
FTS5Pattern(matchingPhrase: "foo bar")?.rawPattern        // "foo bar"
FTS5Pattern(matchingPrefixPhrase: "foo bar")?.rawPattern  // ^"foo bar"

FTS5Pattern(matchingAnyTokenIn: "foo bar*")?.rawPattern   // foo OR bar
FTS5Pattern(matchingAllTokensIn: "foo bar*")?.rawPattern  // foo bar
FTS5Pattern(matchingPhrase: "foo bar*")?.rawPattern       // "foo bar"
FTS5Pattern(matchingPrefixPhrase: "foo bar*")?.rawPattern // ^"foo bar"
```

When you want to build a pattern that uses the `*` wildcard, you must use a raw pattern.

When you want to build such a pattern from user input, without requiring your users to learn the FTS5 query grammar, you can run `FTS5Pattern(matchingAllTokensIn:)`, split the returned raw pattern on the space character, add `*` where you want, and run `Database.makeFTS5Pattern(rawPattern:forTable:)` on the resulting string.

**Comment by groue** [comment]
Closing due to lack of activity
